### ISSUE TITLE
parse error results in XMPMeta NULL

### ISSUE THREAD

**Issue by michaelosthege** [drewnoakes/metadata-extractor-dotnet#51]
I have a picture that was face-tagged by Windows Live Photo Gallery. I would like to extract the face tag.. (Windows 8.1 app)

```
IEnumerable<Directory> directories = ImageMetadataReader.ReadMetadata(await imageFile.OpenStreamForReadAsync());
XmpDirectory xmpDir = (from Directory d in directories where d.Name == "XMP" select d).FirstOrDefault() as XmpDirectory;

```
At this point xmpDir.XmpMeta is null and the Errors enumeration contains "Error processing XMP data: Nested content not allowed with rdf:resource or property attributes".

I have attached the file so you can reproduce.

With http://www.exifviewer.org/ the XMP is correctly extracted:

```
Xmp.xmp.CreatorTool                          XmpText    37  Microsoft Photo Gallery 16.4.3528.331
Xmp.MP.RegionInfo                            XmpText     0  type="Struct"
Xmp.MP.RegionInfo/MPRI:Regions               XmpText     0  type="Bag"
Xmp.MP.RegionInfo/MPRI:Regions[1]            XmpText     0  type="Struct"
Xmp.MP.RegionInfo/MPRI:Regions[1]/MPReg:Rectangle XmpText    38  0.403125, 0.330598, 0.192188, 0.288394
Xmp.MP.RegionInfo/MPRI:Regions[1]/MPReg:PersonDisplayName XmpText    11  Bill Murray
Xmp.xmpMM.InstanceID                         XmpText    41  uuid:faf5bdd5-ba3d-11da-ad31-d33d75182f1b
```

Any clues at which level this issue is born?
Is there another way to get the raw XMP XML?

cheers
michael

![bill murray](https://cloud.githubusercontent.com/assets/5894642/15718155/5b90fb60-282a-11e6-873f-602cd2751696.jpg)

**Comment by michaelosthege** [comment]
I traced it back to XmpCore.Impl.ParseRdf.Rdf_EmptyPropertyElement. It throws "Nested content not allowed with rdf:resource or property attributes" because it doesn't like the structure of the Xmp xml:

```
<?xpacket begin='﻿' id='W5M0MpCehiHzreSzNTczkc9d'?>
<x:xmpmeta xmlns:x="adobe:ns:meta/">
  <rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
    <rdf:Description rdf:about="uuid:faf5bdd5-ba3d-11da-ad31-d33d75182f1b" xmlns:xmp="http://ns.adobe.com/xap/1.0/">
      <xmp:CreatorTool>Microsoft Photo Gallery 16.4.3528.331</xmp:CreatorTool>
    </rdf:Description>
    <rdf:Description rdf:about="uuid:faf5bdd5-ba3d-11da-ad31-d33d75182f1b" xmlns:MP="http://ns.microsoft.com/photo/1.2/">
      <MP:RegionInfo>
        <rdf:Description xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
          <MPRI:Regions xmlns:MPRI="http://ns.microsoft.com/photo/1.2/t/RegionInfo#">
            <rdf:Bag xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
              <rdf:li>
                <rdf:Description xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
                  <MPReg:Rectangle xmlns:MPReg="http://ns.microsoft.com/photo/1.2/t/Region#">0.403125, 0.330598, 0.192188, 0.288394</MPReg:Rectangle>
                  <MPReg:PersonDisplayName xmlns:MPReg="http://ns.microsoft.com/photo/1.2/t/Region#">Bill Murray</MPReg:PersonDisplayName>
                </rdf:Description>
              </rdf:li>
            </rdf:Bag>
          </MPRI:Regions>
        </rdf:Description>
			</MP:RegionInfo>
    </rdf:Description>
  </rdf:RDF>
</x:xmpmeta>
```

Any ideas why this XMP (generated by Windows Live Photo Gallery) is not accepted?

**Comment by michaelosthege** [comment]
I made myself a workaround by **creating a "Root" property of type XDocument on the XmpDirectory**.
In **XmpReader.Extract**, I then pulled the loading behaviour into two steps:
```
// first extract the XML
directory.SetRootDocument(XmpMetaFactory.ExtractRootFromBuffer(xmpBytes));
// now try to parse it
var xmpMeta = XmpMetaFactory.ParseFromRoot(directory.Root);
```
And in XMPMetaFactory:
```
public static XDocument ExtractRootFromBuffer(byte[] buffer, ParseOptions options = null)
{
      return XmpMetaParser.Extract(buffer, options);
}
public static IXmpMeta ParseFromRoot(XDocument root, ParseOptions options = null)
{
    return XmpMetaParser.Parse(root, options);
}
```

**Comment by kwhopper** [comment]
This should be fixed in the XmpCore project once it's merged. It's likely other users have had similar problems with other xml's.

I thought there was maybe a way to get the XMP xml without going through the XMP reader, but now I can't remember. Maybe @drewnoakes has some ideas.

**Comment by drewnoakes** [comment]
That PR is now merged. I'll update the dependency soon. Am out in the country for the rest of the day.

**Comment by drewnoakes** [comment]
So I actually updated the dependency on Jun 14. Have you had a chance to test whether this issue is fixed for you now?

**Comment by drewnoakes** [comment]
Running your image through the latest version of the library doesn't produce the error, so I'm thinking this is fixed. Let me know if you disagree.

```text
FILE: metadata-extractor-dotnet issue 51.jpg

[JPEG - 0xfffffffd] Compression Type = Baseline
[JPEG - 0x0000] Data Precision = 8 bits
[JPEG - 0x0001] Image Height = 853 pixels
[JPEG - 0x0003] Image Width = 1280 pixels
[JPEG - 0x0005] Number of Components = 3
[JPEG - 0x0006] Component 1 = Y component: Quantization table 0, Sampling factors 2 horiz/1 vert
[JPEG - 0x0007] Component 2 = Cb component: Quantization table 1, Sampling factors 1 horiz/1 vert
[JPEG - 0x0008] Component 3 = Cr component: Quantization table 1, Sampling factors 1 horiz/1 vert

[JpegComment - 0x0000] JPEG Comment = File source: http://commons.wikimedia.org/wiki/File:Bill_Murray-3_Get_Low_TIFF09.jpg

[JFIF - 0x0005] Version = 1.1
[JFIF - 0x0007] Resolution Units = inch
[JFIF - 0x0008] X Resolution = 72 dots
[JFIF - 0x000a] Y Resolution = 72 dots
[JFIF - 0x000c] Thumbnail Width Pixels = 0
[JFIF - 0x000d] Thumbnail Height Pixels = 0

[Exif IFD0 - 0x0132] Date/Time = 2016:06:01 18:42:53
[Exif IFD0 - 0xea1c] Padding = [2060 values]

[Exif SubIFD - 0xea1c] Padding = [2060 values]

[Exif Thumbnail - 0x0103] Compression = JPEG (old-style)
[Exif Thumbnail - 0x011a] X Resolution = 72 dots per inch
[Exif Thumbnail - 0x011b] Y Resolution = 72 dots per inch
[Exif Thumbnail - 0x0128] Resolution Unit = Inch
[Exif Thumbnail - 0x0201] Thumbnail Offset = 4304 bytes
[Exif Thumbnail - 0x0202] Thumbnail Length = 7905 bytes

[XMP - 0xffff] XMP Value Count = 7
[XMP - 0x0203] Creator Tool = Microsoft Photo Gallery 16.4.3528.331

[XMPMeta - http://ns.adobe.com/xap/1.0/]  = 
[XMPMeta - http://ns.adobe.com/xap/1.0/] xmp:CreatorTool = Microsoft Photo Gallery 16.4.3528.331
[XMPMeta - http://ns.microsoft.com/photo/1.2/]  = 
[XMPMeta - http://ns.microsoft.com/photo/1.2/] MP:RegionInfo = 
[XMPMeta - http://ns.microsoft.com/photo/1.2/t/RegionInfo#] MP:RegionInfo/MPRI:Regions = 
[XMPMeta - ] MP:RegionInfo/MPRI:Regions[1] = 
[XMPMeta - http://ns.microsoft.com/photo/1.2/t/Region#] MP:RegionInfo/MPRI:Regions[1]/MPReg:Rectangle = 0.403125, 0.330598, 0.192188, 0.288394
[XMPMeta - http://ns.microsoft.com/photo/1.2/t/Region#] MP:RegionInfo/MPRI:Regions[1]/MPReg:PersonDisplayName = Bill Murray
[XMPMeta - http://ns.adobe.com/xap/1.0/mm/]  = 
[XMPMeta - http://ns.adobe.com/xap/1.0/mm/] xmpMM:InstanceID = uuid:faf5bdd5-ba3d-11da-ad31-d33d75182f1b

[File - 0x0001] File Name = metadata-extractor-dotnet issue 51.jpg
[File - 0x0002] File Size = 132596 bytes
[File - 0x0003] File Modified Date = <omitted for regression testing as checkout dependent>

- JPEG
- JpegComment
- JFIF
- Exif IFD0
    - Exif SubIFD
    - Exif Thumbnail
- XMP
- File

Generated using metadata-extractor
https://drewnoakes.com/code/exif/
```

**Comment by michaelosthege** [comment]
I'm actually using a forked version because I had to add multiple things in order to enable writing of XMP (writing the Xml part to be more precise) which I actually got to work. You merged the exact same code that I use locally, so I assume the issue is indeed resolved.

Concerning the XMP writing, I intend to do the diff and open a PR as soon as I'm back in the country...
I'll also test it on the release branch then.
cheers
